internal/controller/customreplicaset_controller.go

```diff
@@ -336,215 +359,223 @@ func (r *CustomReplicaSetReconciler) findChildPods(ctx context.Context, req ctrl
-			return err
-		}
-	}
-	return nil
-}
-
-func (r *CustomReplicaSetReconciler) createPods(ctx context.Context, totalPods, stdPods, upgPods int, cr *customreplicasetv1.CustomReplicaSet, log logr.Logger) error {
-	for totalPods < int(cr.Spec.Replicas) {
-		var newPod *corev1.Pod
-		// Create any missing upgraded pods first
-		if upgPods < int(cr.Spec.Partition) {
-			newPod = r.newPodForCR(cr, true)
-			upgPods++
-		} else {
-			newPod = r.newPodForCR(cr, false)
-			stdPods++
-		}
-
-		if err := r.Create(ctx, newPod); err != nil {
-			log.Error(err, "Unable to create new pod")
-			return err
+func (r *CustomReplicaSetReconciler) managePods(ctx context.Context, crs *customreplicasetv1.CustomReplicaSet, childPods corev1.PodList, podsPerRevision map[int][]*corev1.Pod, totalAvailPods int, latestRevision *v1.ControllerRevision) error {
+	latestPods, ok := podsPerRevision[int(latestRevision.Revision)]
+	if !ok {
+		fmt.Println("No pods found of latest revision")
+		latestPods = []*corev1.Pod{}
+	}
+	if totalAvailPods < int(crs.Spec.Replicas) {
+		podsToCreate := int(crs.Spec.Replicas) - totalAvailPods
+		newPodsToCreate := int(crs.Spec.Partition) - len(latestPods)
+		oldPodsToCreate := podsToCreate - newPodsToCreate
+		if newPodsToCreate > 0 {
+			if err := r.createPods(ctx, crs, newPodsToCreate, podsPerRevision, latestRevision); err != nil {
+				fmt.Println("Failed to create pods of latest version", err)
+				return err
+			}
 		}
-		fmt.Println("Created Pod")
-		totalPods++
-	}
-	return nil
-}
 
-func (r *CustomReplicaSetReconciler) deletePods(ctx context.Context, totalPods, stdPods, upgPods int, cr *customreplicasetv1.CustomReplicaSet, childPods corev1.PodList, log logr.Logger) error {
-	podsToDelete := totalPods - int(cr.Spec.Replicas)
-	upgPodsToDelete := upgPods - int(cr.Spec.Partition)
-	stdPodsToDelete := podsToDelete - upgPodsToDelete
-
-	for _, pod := range childPods.Items {
-		if pod.Status.Phase == corev1.PodRunning || pod.Status.Phase == corev1.PodPending {
-			if strings.Contains(pod.ObjectMeta.Name, "upgraded") {
-				if upgPodsToDelete > 0 {
-					upgPodsToDelete--
-				} else {
-					continue
-				}
-			} else {
-				if stdPodsToDelete > 0 {
-					stdPodsToDelete--
-				} else {
-					continue
-				}
+		if oldPodsToCreate > 0 {
+			oldRevision, err := r.getControllerRevisionExcludingRevision(ctx, crs, latestRevision.Revision)
+			if err != nil {
+				fmt.Println("Failed to get a controller revision excluding the latest one")
+				return err
 			}
-
-			if err := r.Delete(ctx, &pod); err != nil {
-				log.Error(err, "Unable to delete pod")
+			if err := r.createPods(ctx, crs, oldPodsToCreate, podsPerRevision, oldRevision); err != nil {
+				fmt.Println("Failed to create pods", err)
+				return err
+			}
+		}
+	} else if totalAvailPods > int(crs.Spec.Replicas) {
+		podsToDelete := totalAvailPods - int(crs.Spec.Replicas)
+		newPodsToDelete := len(latestPods) - int(crs.Spec.Partition)
```

Comment:
what if newPodsToDelete is a negative value? `oldPodsToDelete := podsToDelete - newPodsToDelete` is wrong.

Reply:
Yes right I put a floor cap of 0 to ensure this doesnt happen